src/pybind/mgr/mgr_module.py

```diff
@@ -330,6 +330,70 @@ def device_class_counts(self) -> Dict[str, int]:
 HandlerFuncType = Callable[..., Tuple[int, str, str]]
 
 
+def _extract_target_func(
+    f: HandlerFuncType
+) -> Tuple[HandlerFuncType, Dict[str, Any]]:
+    """In order to interoperate with other decorated functions,
+    we need to find the original function which will provide
+    the main set of arguments. While we descend through the
+    stack of wrapped functions, gather additional arguments
+    the decorators may want to provide.
+    """
+    wrapped = getattr(f, "__wrapped__", None)
+    if not wrapped:
+        return f, {}
+    extra_args = {}
+    while wrapped is not None:
+        extra_args.update(getattr(f, "extra_args", {}))
+        f = wrapped
+        wrapped = getattr(f, "__wrapped__", None)
+    return f, extra_args
+
+
+def _load_func_metadata(
```

Comment:
That's interesting as I've always considered staticmethod a "code smell". I don't feel as if I should go deeply into the reasons why I'll link to [this blog post](https://www.seporaitis.net/posts/2020/05/05/python-staticmethod-usage/) I found in a quick search - while I don't share 100% of my opinions of it, it overlaps well enough for this part of the discussion.

Moving on, I'm not married to the idea of having this as a private module level function. And as you point out it consumes KNOWN_ARGS. So, based on your comment I think would you agree to me changing it to a private classmethod?

Finally, I agree with the general idea of not mixing structural changes and logic changes together. However, I put that at a commit level. Are you arguing for not having structural and logic changes together in a PR even if they're multiple patches/commits?

Reply:
Given that Python didn't support `namespace`s until recently, I guess that `staticmethod` kinda filled that void.

In this very case IMHO this shouldn't be a real `staticmethod` since it's consuming a class variable, so using `classmethod` is a kind way of making dependencies explicit. If we all agree that global variables should be avoided, then we shouldn't access class members from external functions either.

And regarding the private-protected-public topic, since Python has never been very serious about it (unlike other OOP languages, nothing prevents you from accessing a private member, and "_protected" is just a ~joke~ convention that only static checkers enforce) I usually don't spend too much time thinking of it.

I'm just concerned about shifting Python from OOP to a more procedural approach. While files/modules might behave as namespaces/scopes too, a well designed class can be extended by inheritance, something impossible with files/modules.